apiserver/common/tools.go

```diff
@@ -277,48 +277,84 @@ func (f *ToolsFinder) findTools(args params.FindToolsParams) (coretools.List, er
-		args.OSType = ""
-		compatibleMatch = true
-	}
-
-	storageList, err := f.matchingStorageTools(args)
-	if err != nil && err != coretools.ErrNoMatches {
-		return nil, err
-	}
+	// The following functions convert the result tools list to the required format.
 
 	// For a given list of tools, return those which match the required
-	// os type based on an exact os type match or series match.
-	compatibleTools := func(tools coretools.List) coretools.List {
+	// os type. If the tools is for a series, first convert the series into
+	// the corresponding os type before doing the match.
+	// eg if ubuntu tools are requested, and we have bionic tools, say they're ubuntu.
+	wantedOSType := args.OSType
+	compatibleOSTypeTools := func() {
 		added := make(map[version.Binary]bool)
 		var matched coretools.List
-		for _, t := range tools {
+		for _, t := range result {
 			converted := *t
-			osTypeName, _ := coreseries.GetOSFromSeries(t.Version.Release)
-			if osTypeName != coreos.Unknown {
+			// t might be for a series so convert to an os type.
+			if !coreos.IsValidOSTypeName(t.Version.Release) {
+				osTypeName, err := coreseries.GetOSFromSeries(t.Version.Release)
+				if err != nil {
+					continue
+				}
 				converted.Version.Release = strings.ToLower(osTypeName.String())
 			}
+			if converted.Version.Release != wantedOSType {
+				continue
+			}
 			if added[converted.Version] {
 				continue
 			}
-			if converted.Version.Release == wantedOSType || wantedOSType == "" {
-				matched = append(matched, &converted)
-				added[converted.Version] = true
+			matched = append(matched, &converted)
+			added[converted.Version] = true
+		}
+		result = matched
+	}
+
+	// For a given list of tools, return those which match the required series.
+	// If the tools is for an os type, match on the os type of the series.
+	// The returned tools are converted to the requested series.
+	// eg if bionic tools are requested, we return bionic if we have then
+	// or convert ubuntu tools to say they're bionic.
+	compatibleSeriesTools := func() {
+		osType := coreseries.DefaultOSTypeNameFromSeries(args.Series)
+		added := make(map[version.Binary]bool)
+		var matched coretools.List
+		for _, t := range result {
+			converted := *t
+			if coreos.IsValidOSTypeName(t.Version.Release) {
+				if osType != t.Version.Release {
+					continue
+				}
+				converted.Version.Release = args.Series
+			} else if args.Series != t.Version.Release {
+				continue
+			}
+			if added[converted.Version] {
+				continue
 			}
+			matched = append(matched, &converted)
+			added[converted.Version] = true
 		}
-		return matched
+		result = matched
 	}
 
-	if compatibleMatch {
-		storageList = compatibleTools(storageList)
+	// TODO(juju4) - remove this logic
+	// Set up the final metadata conversion if old tools are requested.
+	if args.Number.Major == 2 && args.Number.Minor <= 8 && (args.OSType != "" || args.Series != "") {
+		if args.OSType != "" {
```

Comment:
This couldn't happen though right? Since `args.OSType` was added in the 2.9 series, and since the agent will not be a later version than the controller, we don't even need this.

Reply:
that code was added for CLI client compatibility